### PR TITLE
Attr

### PR DESCRIPTION
**Pull request by mhevery**

### CODE REVIEW

File: src/iv/instructions.ts [hunk 1]
```diff
@@ -181,6 +187,28 @@ export function elementEnd(template?: Template<any>) {
   }
 }
 
+export function elementAttribute(attrIndex: number, attrName: string, values: any[]): void {
+  const ivElement = cursor as IvElement;
+  const state = ivElement.value || (ivElement.value = []);
+  if (creationMode) {
+    state.push(values.slice()); // make a copy of the array.
+    for(let i=0; i< values.length; i++) {
+      values[i] = stringify(values[i]);
```
Comment (pkozlowski-opensource) [inline src/iv/instructions.ts]: Here we would store a stringified value while using non-stringified value for comparison later on. As I understand we want to store non-stringified value in "old values" - otherwise we would have to stringify all the values in the update mode (potentially unnecessary string creation = memory pressure?)

File: src/iv/instructions.ts [hunk 2]
```diff
@@ -181,6 +187,28 @@ export function elementEnd(template?: Template<any>) {
   }
 }
 
+export function elementAttribute(attrIndex: number, attrName: string, values: any[]): void {
+  const ivElement = cursor as IvElement;
+  const state = ivElement.value || (ivElement.value = []);
+  if (creationMode) {
+    state.push(values.slice()); // make a copy of the array.
```
Comment (pkozlowski-opensource) [inline src/iv/instructions.ts]: If we don't store stringified values we don't need to make a copy of this array here, right?

File: src/iv/instructions.ts (hunk 1 above)
Comment (pkozlowski-opensource) [reply src/iv/instructions.ts]: Oh, scratch this, I see what is going on here! Sorry for the noise!

**Comment by pkozlowski-opensource** [comment]
@mhevery it just downed on me that we, in reality, don't have the case of interpolation for attributes (?). In my mind we've got the following cases:

- `foo="static value"`
- `[attr.foo]`="exp" - here `exp` can't have multiple values / interpolation
- `[foo]="{{exp1}} static {{exp2"}` - here we've got interpolation but this is a _property_ binding 

The only case where we could have interpolation for attributes would be for creation-time bindings (not sure what the syntax for it would be) with the generated code being sth like `t(cm && (exp1 + " static " + exp2)` - here we need to concatenate a string for the creation mode (need to set it as an attribute anyway) and won't have string concatenation for update (there is no update).

**Comment by pkozlowski-opensource** [comment]
tl;dr; is that for _attributes_ it is enough to keep just one value (as old value). I think I might have confused you by saying at one point that we need interpolation for attributes, sorry for this. But it looks like b05ba09 should not be merged (still we need a similar solution for properties / directive inputs)

**Comment by mhevery** [comment]
@pkozlowski-opensource 

> `[foo]="{{exp1}} static {{exp2"}` - here we've got interpolation but this is a property binding

The above is not a thing but this is:  `foo="{{exp1}} static {{exp2"}` which is what is implemented by this PR. The only other place where we have interpolation is in the text nodes and i think we should do something similar.

The reason I have `a0-8` is that I don't want to create garbage every time CD happens. We need to store the individual items rather than concatenated string or we will have GC pressure.

Do you have a different idea how this could be implemented without GC pressure? `a` is the generic version and that one will allocate an array every time it runs.

In practice there are only few bindings usually, so everything but `a0-3` will most likely be tree shaken away.

**Comment by pkozlowski-opensource** [comment]
@mhevery thnx for your latest [comment](https://github.com/mhevery/ngiv/pull/9#issuecomment-335132934) - it made me realise that there are different optimization concerns here (generated code size, engine code size, memory pressure, pure runtime perf) and it looks like we are focusing on different aspects.

I totally get what you are saying about unnecessary object creation if we use one of the following forms:
* `t("static " + exp1 + "more static" + exp2 )` (unnecessary string creation)
* `t(["static ",  exp1, "more static", exp2])` (unnecessary array creation) 

> Do you have a different idea how this could be implemented without GC pressure? a is the generic version and that one will allocate an array every time it runs.

In this respect `t0, t1, ...` make total sense. But I wanted to check if it would be possible to use `t("static ",  exp1, "more static", exp2 )` and access the arguments array. I know that accessing `arguments` was a big no-no in Crankshaft but this is no longer a deopt reason in Turbofan. Unless accessing `arguments` will allocate an array - if this is the case than indeed I don't see any other choice as going with `t0, t1, ...`.

Now, my **biggest** concern is that if we apply the `t0, t1, ...` pattern across the board we are going to end up with lots of functions in the engine. 

> The only other place where we have interpolation is in the text nodes and i think we should do something similar.

Actually, I believe that there are many more places where interpolations can by used:
* text nodes: `<span>before {{exp}} after</span>`
* property:  <span title="before {{exp}} after">
* attribute:  <span attr.title="before {{exp}} after">
* class special case: <span class.title="before {{exp}} after">
* directive input:  <span ngbTooltip="before {{exp}} after">

Given all those cases I wouldn't like to duplicate the same code for each and every place where interpolation is need. I think that we need a generic `checkInterpolation` function for which I've got a _very_ dirty prototype here: https://github.com/pkozlowski-opensource/ngiv-interpolations/blob/master/test/interpolations.spec.ts#L59-L79. I will clean it up tomorrow and run some benchmarks on it.

To sum up:
* I'm concerned with the code addition to the engine so toying with a generic `checkInterpolation` function;
* I totally get your memory pressure concerns, will check if we can have only dynamic version by accessing `arguments`;
* we need to run benchmarks to see the actual impact of different approaches to data structures / generated code.

**Comment by mhevery** [comment]
> and access the arguments array. I know that accessing arguments was a big no-no in Crankshaft but this is no longer a deopt reason in Turbofan. Unless accessing arguments will allocate an array - if this is the case than indeed I don't see any other choice as going with t0, t1, ....

Accessing `arguments`:
1) creates an object (gc preasure)
2) is not allowed in 'strict' mode.
3) even if you ignore 1&2 you now haw to have a function which will do something for 1, 2, 3, ... arguments, hence it is either generic (arrays/loops) or specific (if statements) and now we can't tree shake the ones which  we don't need for the app.

> Now, my biggest concern is that if we apply the t0, t1, ... pattern across the board we are going to end up with lots of functions in the engine.

The only place where this needs to be applied is with 1) attributes and 2) text nodes. No other place.

property, attribute, class special case, directive input: do not have this issue. for `class="foo {{bar}}"` should not be supported and instead should be `class="foo" [class.bar]=exp`

> I think that we need a generic checkInterpolation function for which I've got a very dirty prototype here: https://github.com/pkozlowski-opensource/ngiv-interpolations/blob/master/test/interpolations.spec.ts#L59-L79.

all up for a generic interpolation function. However your prototype has array GC creation pressure.

If you could write some benchmarks for this that would be great.
